gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/TokenStateDatabase.java

```diff
@@ -0,0 +1,145 @@
+  }
+
+  // This needs to be done in batches as many DB vendors have limitations on the number of items within the 'IN' clause
+  boolean removeTokens(Set<String> tokenIds) throws SQLException {
+    int removed = 0;
+    if (tokenIds.size() <= DELETE_BATCH_SIZE) {
+      removed += doRemoveTokens(tokenIds);
+    } else {
+      Set<String> tokenIdBatch = new HashSet<>();
+      for (String tokenId : tokenIds) {
+        tokenIdBatch.add(tokenId);
+        if (tokenIdBatch.size() == DELETE_BATCH_SIZE) {
+          removed += doRemoveTokens(tokenIdBatch);
+          tokenIdBatch.clear();
+        }
+      }
+      // one more round of removal if the last batch has less items than the configured batch size
+      removed += doRemoveTokens(tokenIdBatch);
+    }
+    return removed == tokenIds.size();
+  }
+
+  private int doRemoveTokens(Set<String> tokenIds) throws SQLException {
+    final StringBuilder statementPostFixBuilder = new StringBuilder(REMOVE_TOKENS_SQL_PREFIX);
+    for (int i = 0; i < tokenIds.size(); i++) {
+      if (statementPostFixBuilder.length() > REMOVE_TOKENS_SQL_PREFIX.length()) {
+        statementPostFixBuilder.append(", ");
+      }
+      statementPostFixBuilder.append('?');
+    }
+    statementPostFixBuilder.append(')');
+    try (Connection connection = dataSource.getConnection(); PreparedStatement removeTokensStatement = connection.prepareStatement(statementPostFixBuilder.toString())) {
+      int i = 0;
+      for (String tokenId : tokenIds) {
+        removeTokensStatement.setString(++i, tokenId);
+      }
+      return removeTokensStatement.executeUpdate();
+    }
+  }
+
+  long getTokenExpiration(String tokenId) throws SQLException {
+    try (Connection connection = dataSource.getConnection(); PreparedStatement getTokenExpirationStatement = connection.prepareStatement(GET_TOKEN_EXPIRATION_SQL)) {
+      getTokenExpirationStatement.setString(1, tokenId);
+      try (ResultSet rs = getTokenExpirationStatement.executeQuery()) {
+        return rs.next() ? rs.getLong(1) : 0;
+      }
+    }
+  }
+
+  boolean updateExpiration(final String tokenId, long expiration) throws SQLException {
+    try (Connection connection = dataSource.getConnection(); PreparedStatement updateTokenExpirationStatement = connection.prepareStatement(UPDATE_TOKEN_EXPIRATION_SQL)) {
+      updateTokenExpirationStatement.setLong(1, expiration);
+      updateTokenExpirationStatement.setString(2, tokenId);
+      return updateTokenExpirationStatement.executeUpdate() == 1;
+    }
+  }
+
+  long getMaxLifetime(String tokenId) throws SQLException {
+    try (Connection connection = dataSource.getConnection(); PreparedStatement getMaxLifetimeStatement = connection.prepareStatement(GET_MAX_LIFETIME_SQL)) {
+      getMaxLifetimeStatement.setString(1, tokenId);
+      try (ResultSet rs = getMaxLifetimeStatement.executeQuery()) {
+        return rs.next() ? rs.getLong(1) : -1;
+      }
+    }
+  }
+
+  List<String> getTokenIds() throws SQLException {
```

Comment:
This looks risky :) we might overwhelm Knox depending on number of tokens, could end up DOSing

Reply:
Two notes here:
- It's not part of any of the public API invocations so there is no risk here in terms of DOS
- Even if we are talking about a large number of tokens, getting them from the database should not be a problem

I'm going to run our own performance testing tool with this implementation to see actual numbers (e.g. having 1 million tokens in the DB) but I doubt that querying even millions of tokens from a DB causes a problem.

Reply:
cool! future devs might use this method accidentally and could cause issues but if it not public should be ok.

Reply:
With the new implementation, this is not a problem anymore (I did not even overwrite this method)